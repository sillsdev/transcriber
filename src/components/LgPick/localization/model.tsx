// WARNING: This file is generated using ToModel.xsl. Changes made here may be lost.
import * as Localize from 'react-localization';

export interface ILanguagePickerStrings extends Localize.LocalizedStringsMethods {
    "font": string;
    "script": string;
    "language": string;
    "selectLanguage": string;
    "findALanguage": string;
    "codeExplained": string;
    "subtags": string;
    "details": string;
    "languageOf": string;
    "inScript": string;
    "select": string;
    "cancel": string;
};


export interface ILocalizedStrings {
	loaded: boolean;
	lang: string;
	languagePicker: ILanguagePickerStrings;
	[key: string]: any;
};
