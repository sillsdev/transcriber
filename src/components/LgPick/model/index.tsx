export * from "../langPicker/types";
export * from "../localization/model";
