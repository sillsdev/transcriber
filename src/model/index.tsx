export * from './organization';
export * from './organizationMembership';
export * from './group';
export * from './groupMembership';
export * from './integration';
export * from './invitation';
export * from './mediaFile';
export * from './project';
export * from './projectType';
export * from './plan';
export * from './planType';
export * from './section';
export * from './passage';
export * from './passageSection';
export * from './user';
export * from './role';
export * from './userPassage';
export * from './activityState';
export * from './state';
export * from '../store/book/types';
export * from '../store/localization/types';
export * from '../store/localization/model';
export * from '../store/orbit/types';
export * from '../store/upload/types';
export * from '../store/context/types';
export * from '../store/langPicker/types';
export * from '../store/media/types';
export * from '../store/paratext/types';
