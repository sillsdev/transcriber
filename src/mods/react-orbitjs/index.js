Object.defineProperty(exports, '__esModule', {
  value: true,
});
exports.withData = exports.DataProvider = undefined;

var _DataProvider = require('./components/DataProvider');

var _DataProvider2 = _interopRequireDefault(_DataProvider);

var _withData = require('./components/withData');

var _withData2 = _interopRequireDefault(_withData);

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

exports.DataProvider = _DataProvider2.default;
exports.withData = _withData2.default;
