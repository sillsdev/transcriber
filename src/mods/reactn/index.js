var components_1 = require('./components');
var add_callback_1 = require('./add-callback');
var add_reducer_1 = require('./add-reducer');
var add_reducers_1 = require('./add-reducers');
var create_provider_1 = require('./create-provider');
var decorator_1 = require('./decorator');
var default_global_state_manager_1 = require('./default-global-state-manager');
var get_dispatch_1 = require('./get-dispatch');
var get_global_1 = require('./get-global');
var remove_callback_1 = require('./remove-callback');
var reset_global_1 = require('./reset-global');
var set_global_1 = require('./set-global');
var use_dispatch_1 = require('./use-dispatch');
var use_global_1 = require('./use-global');
var with_global_1 = require('./with-global');
var with_init_1 = require('./with-init');
var React = require('react');
module.exports = Object.assign(decorator_1.default, React, {
  addCallback: add_callback_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  addReducer: add_reducer_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  addReducers: add_reducers_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  Component: components_1.ReactNComponent,
  createProvider: create_provider_1.default,
  default: decorator_1.default,
  getDispatch: get_dispatch_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  getGlobal: get_global_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  PureComponent: components_1.ReactNPureComponent,
  removeCallback: remove_callback_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  resetGlobal: reset_global_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  setGlobal: set_global_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  useDispatch: use_dispatch_1.default.bind(null, null),
  useGlobal: use_global_1.default.bind(null, null),
  withGlobal: with_global_1.default.bind(null, null),
  withInit: with_init_1.default,
});
