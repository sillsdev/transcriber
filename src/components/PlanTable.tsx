import React, { useState, useEffect } from 'react';
import { useGlobal } from 'reactn';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { IState, Plan, PlanType, Section, IPlanTableStrings } from '../model';
import localStrings from '../selector/localize';
import { withData } from 'react-orbitjs';
import { QueryBuilder, TransformBuilder } from '@orbit/data';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import { Button } from '@material-ui/core';
import AddIcon from '@material-ui/icons/Add';
import DeleteIcon from '@material-ui/icons/Delete';
import EditIcon from '@material-ui/icons/Edit';
import IconButton from '@material-ui/core/IconButton';
import FilterIcon from '@material-ui/icons/FilterList';
import SelectAllIcon from '@material-ui/icons/SelectAll';
import { Table } from '@devexpress/dx-react-grid-material-ui';
import ShapingTable from './ShapingTable';
import PlanAdd from './PlanAdd';
import SnackBar from './SnackBar';
import Confirm from './AlertDialog';
import Related from '../utils/related';
import { numCompare } from '../utils/sort';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      width: '100%',
    },
    container: {
      display: 'flex',
      justifyContent: 'center',
    },
    paper: theme.mixins.gutters({
      display: 'flex',
      flexDirection: 'column',
      alignContent: 'center',
    }),
    grow: {
      flexGrow: 1,
    },
    dialogHeader: theme.mixins.gutters({
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'center',
    }),
    editIcon: {
      fontSize: 16,
    },
    link: {},
    actionIcon: {},
    button: {},
    icon: {},
    buttonIcon: {
      marginLeft: theme.spacing(1),
    },
  })
);

interface IRow {
  name: string;
  planType: string;
  sections: string;
  action: string;
}

interface IStateProps {
  t: IPlanTableStrings;
  tableLoad: string[];
}

interface IRecordProps {
  plans: Array<Plan>;
  planTypes: Array<PlanType>;
  sections: Array<Section>;
}

interface IProps extends IStateProps, IRecordProps {
  displaySet: (type: string) => any;
}

export function PlanTable(props: IProps) {
  const { plans, planTypes, sections, t, displaySet, tableLoad } = props;
  const classes = useStyles();
  const [schema] = useGlobal('schema');
  const [memory] = useGlobal('memory');
  const [projRole] = useGlobal('projRole');
  const [project] = useGlobal('project');
  const [columns] = useState([
    { name: 'name', title: t.name },
    { name: 'planType', title: t.type },
    { name: 'sections', title: t.sections },
    { name: 'action', title: projRole === 'admin' ? t.action : '\u00A0' },
  ]);
  const [columnWidth] = useState([
    { columnName: 'name', width: 300 },
    { columnName: 'planType', width: 170 },
    { columnName: 'sections', width: 100 },
    { columnName: 'action', width: 150 },
  ]);
  const columnSorting = [{ columnName: 'sections', compare: numCompare }];
  const sortingEnabled = [{ columnName: 'action', sortingEnabled: false }];
  const filteringEnabled = [{ columnName: 'action', filteringEnabled: false }];

  const numCols = ['sections'];
  const [rows, setRows] = useState(Array<IRow>());
  const [filter, setFilter] = useState(false);
  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
  const [view, setView] = useState('');
  const [deleteItem, setDeleteItem] = useState('');
  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
  const [_plan, setPlan] = useGlobal('plan');
  const [dialogVisible, setDialogVisible] = useState(false);
  const [dialogData, setDialogData] = useState(null as Plan | null);
  const [message, setMessage] = useState(<></>);
  const [loading, setLoading] = useState(false);

  const handleMessageReset = () => {
    setMessage(<></>);
  };
  const handleDelete = (value: string) => () => setDeleteItem(value);
  const handleDeleteConfirmed = () => {
    memory.update((t: TransformBuilder) =>
      t.removeRecord({
        type: 'plan',
        id: deleteItem,
      })
    );
  };
  const handleDeleteRefused = () => {
    setDeleteItem('');
  };
  const handleAdd = () => {
    setDialogData(null);
    setDialogVisible(true);
  };
  const handleAddMethod = async (name: string, planType: string) => {
    setDialogVisible(false);
    let plan: Plan = {
      type: 'plan',
      attributes: {
        name: name,
      },
    } as any;
    schema.initializeRecord(plan);
    await memory.update((t: TransformBuilder) => [
      t.addRecord(plan),
      t.replaceRelatedRecord({ type: 'plan', id: plan.id }, 'plantype', {
        type: 'plantype',
        id: planType,
      }),
      t.replaceRelatedRecord({ type: 'plan', id: plan.id }, 'project', {
        type: 'project',
        id: project,
      }),
    ]);
    handleSelect(plan.id);
  };
  const handleAddCancel = () => {
    setDialogVisible(false);
  };
  const handleEdit = (planId: string) => (e: any) => {
    const planRec = plans.filter(p => p.id === planId);
    setDialogData(planRec && planRec.length === 1 ? planRec[0] : null);
    setDialogVisible(true);
  };
  const handleEditMethod = async (plan: any) => {
    setDialogVisible(false);
    delete plan.relationships;
    await memory.update((t: TransformBuilder) => t.updateRecord(plan));
    await memory.update((t: TransformBuilder) =>
      t.replaceRelatedRecord({ type: 'plan', id: plan.id }, 'plantype', {
        type: 'plantype',
        id: plan.attributes.planType,
      })
    );
  };
  const handleFilter = () => setFilter(!filter);
  const handleSelectEv = (planId: string) => (e: any) => handleSelect(planId);
  const handleSelect = (planId: string) => {
    setPlan(planId);
    const planRec = memory.cache.query((q: QueryBuilder) =>
      q.findRecord({ type: 'plan', id: planId })
    ) as Plan;
    const typeId = Related(planRec, 'plantype');
    const typeRec = memory.cache.query((q: QueryBuilder) =>
      q.findRecord({ type: 'plantype', id: typeId })
    ) as PlanType;
    if (typeRec.attributes && typeRec.attributes.name)
      displaySet(typeRec.attributes.name.toLowerCase());
  };
  const getType = (p: Plan) => {
    const typeId = Related(p, 'plantype');
    const typeRec = planTypes.filter(pt => pt.id === typeId);
    return typeRec && typeRec.length === 1 && typeRec[0].attributes
      ? t.getString(typeRec[0].attributes.name.toLowerCase())
      : '--';
  };
  const sectionCount = (p: Plan) => {
    return sections.filter(s => Related(s, 'plan') === p.id).length.toString();
  };

  useEffect(() => {
    const projectPlans = plans.filter(p => Related(p, 'project') === project);
    setRows(
      projectPlans.map((p: Plan) => {
        return {
          name: p.attributes.name,
          planType: getType(p),
          sections: sectionCount(p),
          action: p.id,
        } as IRow;
      })
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [plans, project]);

  useEffect(() => {
    if (
      tableLoad.length > 0 &&
      (!tableLoad.includes('plantype') || !tableLoad.includes('section')) &&
      !loading
    ) {
      setMessage(<span>{t.loadingTable}</span>);
      setLoading(true);
    } else if (loading) {
      setMessage(<></>);
      setLoading(false);
    }
    /* eslint-disable-next-line react-hooks/exhaustive-deps */
  }, [tableLoad]);

  interface ICell {
    value: string;
    style?: React.CSSProperties;
    row: IRow;
    column: any;
    tableRow: any;
    tableColumn: any;
  }

  const LinkCell = ({ value, style, ...restProps }: ICell) => (
    <Table.Cell {...restProps} style={{ ...style }} value>
      <Button
        key={value}
        aria-label={value}
        color="primary"
        className={classes.link}
        onClick={handleSelectEv(restProps.row.action)}
      >
        {value}
        {/* <EditIcon className={classes.editIcon} /> */}
      </Button>
    </Table.Cell>
  );

  const ActionCell = ({ value, style, ...restProps }: ICell) => (
    <Table.Cell {...restProps} style={{ ...style }} value>
      {projRole === 'admin' && (
        <>
          <IconButton
            id={'edit-' + value}
            key={'edit-' + value}
            aria-label={'edit-' + value}
            color="default"
            className={classes.actionIcon}
            onClick={handleEdit(value)}
          >
            <EditIcon />
          </IconButton>
          <IconButton
            id={'del-' + value}
            key={'del-' + value}
            aria-label={'del-' + value}
            color="default"
            className={classes.actionIcon}
            onClick={handleDelete(value)}
          >
            <DeleteIcon />
          </IconButton>
        </>
      )}
    </Table.Cell>
  );

  const Cell = (props: ICell) => {
    const { column } = props;
    if (column.name === 'name') {
      return <LinkCell {...props} />;
    }
    if (column.name === 'action') {
      return <ActionCell {...props} />;
    }
    return <Table.Cell {...props} />;
  };

  if (view !== '') return <Redirect to={view} />;

  return (
    <div className={classes.root}>
      <div className={classes.container}>
        <div className={classes.paper}>
          <div className={classes.dialogHeader}>
            {projRole === 'admin' && (
              <>
                <Button
                  key="add"
                  aria-label={t.addPlan}
                  variant="outlined"
                  color="primary"
                  className={classes.button}
                  onClick={handleAdd}
                >
                  {t.addPlan}
                  <AddIcon className={classes.buttonIcon} />
                </Button>
              </>
            )}
            <div className={classes.grow}>{'\u00A0'}</div>
            <Button
              key="filter"
              aria-label={t.filter}
              variant="outlined"
              color="primary"
              className={classes.button}
              onClick={handleFilter}
              title={'Show/Hide filter rows'}
            >
              {t.filter}
              {filter ? (
                <SelectAllIcon className={classes.icon} />
              ) : (
                <FilterIcon className={classes.icon} />
              )}
            </Button>
          </div>
          <ShapingTable
            columns={columns}
            columnWidths={columnWidth}
            sortingEnabled={sortingEnabled}
            filteringEnabled={filteringEnabled}
            columnSorting={columnSorting}
            dataCell={Cell}
            sorting={[{ columnName: 'name', direction: 'asc' }]}
            numCols={numCols}
            shaping={filter}
            rows={rows}
          />
        </div>
      </div>
      <PlanAdd
        visible={dialogVisible}
        planIn={dialogData}
        addMethod={handleAddMethod}
        editMethod={handleEditMethod}
        cancelMethod={handleAddCancel}
      />
      {deleteItem !== '' ? (
        <Confirm
          yesResponse={handleDeleteConfirmed}
          noResponse={handleDeleteRefused}
        />
      ) : (
        <></>
      )}
      <SnackBar {...props} message={message} reset={handleMessageReset} />
    </div>
  );
}

const mapStateToProps = (state: IState): IStateProps => ({
  t: localStrings(state, { layout: 'planTable' }),
  tableLoad: state.orbit.tableLoad,
});

const mapRecordsToProps = {
  plans: (q: QueryBuilder) => q.findRecords('plan'),
  planTypes: (q: QueryBuilder) => q.findRecords('plantype'),
  sections: (q: QueryBuilder) => q.findRecords('section'),
};

export default withData(mapRecordsToProps)(
  connect(mapStateToProps)(PlanTable) as any
) as any;
