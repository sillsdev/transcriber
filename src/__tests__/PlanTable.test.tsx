import React from 'react';
import { Router } from 'react-router-dom';
import { setGlobal } from 'reactn';
import { Section } from '../model';
import { DataProvider } from 'react-orbitjs';
import { Provider } from 'react-redux';
import configureStore from '../store';
import { createMuiTheme } from '@material-ui/core';
import Memory from '@orbit/memory';
import { schema, keyMap } from '../schema';
import history from '../history';
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import PlanTable from '../components/PlanTable';

const theme = createMuiTheme({});

const store = configureStore();

const memory = new Memory({ schema, keyMap });

const globals = {
  organization: null,
  orgRole: 'admin',
  project: null,
  projRole: 'admin',
  plan: null,
  tab: 0,
  group: '',
  user: null,
  lang: 'en',
  memory: memory,
  schema: schema,
  keyMap: keyMap,
};
setGlobal(globals);

const tree = (props?: any) => (
  <DataProvider dataStore={memory}>
    <Provider store={store}>
      <Router history={history}>
        <PlanTable {...props} />
      </Router>
    </Provider>
  </DataProvider>
);

const addProjects = async () => {
  const project1 = {
    type: 'project',
    attributes: {
      name: 'Fulfulde',
    },
  } as any;
  schema.initializeRecord(project1);
  setGlobal({ ...globals, project: project1.id });
  await (memory as Memory).update(t => t.addRecord(project1));
  const project2 = {
    type: 'project',
    attributes: {
      name: 'Ewondo',
    },
  } as any;
  schema.initializeRecord(project2);
  await (memory as Memory).update(t => t.addRecord(project2));
  return [project1.id, project2.id];
};

const addPlan = async (project1: string) => {
  const planType = {
    type: 'plantype',
    attributes: {
      name: 'Scripture',
    },
  } as any;
  schema.initializeRecord(planType);
  await (memory as Memory).update(t => t.addRecord(planType));
  const plan = {
    type: 'plan',
    attributes: {
      name: 'Genesis',
    },
  } as any;
  schema.initializeRecord(plan);
  await (memory as Memory).update(t => t.addRecord(plan));
  await (memory as Memory).update(t =>
    t.replaceRelatedRecord({ type: 'plan', id: plan.id }, 'plantype', {
      type: 'plantype',
      id: planType.id,
    })
  );
  await (memory as Memory).update(t =>
    t.replaceRelatedRecord({ type: 'plan', id: plan.id }, 'project', {
      type: 'project',
      id: project1,
    })
  );
  return plan.id;
};

afterEach(cleanup);

test('can render PlanTable snapshot', async () => {
  const { getByText, container } = render(tree());
  await waitForElement(() => getByText(/^Sections$/i));
  expect(container.firstChild).toMatchSnapshot();
});

test('PlanTable displays plans in project name and type', async () => {
  const [project1, project2] = await addProjects();
  await addPlan(project1);

  const projects = await (memory as Memory).query(q =>
    q.findRecords('project')
  );
  const plans = await (memory as Memory).query(q => q.findRecords('plan'));
  const sections = Array<Section>();
  const { getByText, container } = render(tree({ projects, plans, sections }));
  await waitForElement(() => getByText(/^Genesis$/i));
  const tbody = container.querySelector('tbody');
  const button = tbody && tbody.querySelector('button');

  expect(button).not.toBeFalsy;
  expect(getByText(/^Genesis$/i)).toBeTruthy;
  expect(getByText(/^Scripture$/i)).toBeTruthy;
});

test('PlanTable does not display plan of another project', async () => {
  const [project1, project2] = await addProjects();
  await addPlan(project1);

  const projects = await (memory as Memory).query(q =>
    q.findRecords('project')
  );
  const plans = await (memory as Memory).query(q => q.findRecords('plan'));
  const sections = Array<Section>();
  setGlobal({ ...globals, project: project2 });
  const { getByText } = render(tree({ projects, plans, sections }));
  await waitForElement(() => getByText(/^No Data$/i));

  expect(getByText(/^No Data$/i)).toBeTruthy;
});

test('Clicking Add FAB on PlanTable displays dialogue', async () => {
  const [project1, project2] = await addProjects();
  await addPlan(project1);

  const projects = await (memory as Memory).query(q =>
    q.findRecords('project')
  );
  const plans = await (memory as Memory).query(q => q.findRecords('plan'));
  const sections = Array<Section>();
  const { getByText, getByTestId } = render(
    tree({ projects, plans, sections })
  );
  await waitForElement(() => getByText(/^Genesis$/i));
  fireEvent.click(getByText('Add Plan'));
  await waitForElement(() => getByText(/^Type the name/i));
  expect(getByText('Add')).toBeDisabled();
});
