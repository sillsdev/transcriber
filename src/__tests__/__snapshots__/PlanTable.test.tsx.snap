// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render PlanTable snapshot 1`] = `
<div
  class="makeStyles-root-1"
>
  <div
    class="makeStyles-container-2"
  >
    <div
      class="makeStyles-paper-3"
    >
      <div
        class="makeStyles-dialogHeader-5"
      >
        <button
          aria-label="Add Plan"
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary makeStyles-button-9"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add Plan
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-buttonIcon-11"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <div
          class="makeStyles-grow-4"
        >
          Â 
        </div>
        <button
          aria-label="Filter"
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary makeStyles-button-9"
          tabindex="0"
          title="Show/Hide filter rows"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Filter
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-10"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="RootBase-root-52"
      >
        <div
          class="TableContainer-root-53"
        >
          <div>
            <table
              class="MuiTable-root Table-table-54"
              style="min-width: 650px;"
            >
              <colgroup>
                <col
                  style="width: 300px;"
                />
                <col
                  style="width: 100px;"
                />
                <col
                  style="width: 100px;"
                />
                <col
                  style="width: 150px;"
                />
                <col />
              </colgroup>
              <thead
                class="MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head"
                >
                  <th
                    class="MuiTableCell-root TableHeaderCell-cell-67 TableHeaderCell-cellNoUserSelect-71 TableHeaderCell-cellDraggable-72 TableHeaderCell-cellNoWrap-77 MuiTableCell-head"
                    colspan="1"
                    scope="col"
                  >
                    <div
                      class="TableHeaderCell-container-76"
                    >
                      <div
                        class="Content-content-90"
                      >
                        <div
                          class="SortLabel-root-93"
                        >
                          <span
                            aria-disabled="false"
                            class="MuiButtonBase-root MuiTableSortLabel-root SortLabel-sortLabelRoot-95 MuiTableSortLabel-active SortLabel-sortLabelActive-97"
                            role="button"
                            tabindex="0"
                            title="Sort"
                          >
                            <span
                              class="Title-title-111"
                            >
                              Name
                            </span>
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                              focusable="false"
                              role="presentation"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                              />
                            </svg>
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ResizingControl-resizeHandle-112 TableHeaderCell-resizeHandle-68"
                    >
                      <div
                        class="TableHeaderCell-resizeHandleLine-69 ResizingControl-resizeHandleLine-113 ResizingControl-resizeHandleFirstLine-114"
                      />
                      <div
                        class="TableHeaderCell-resizeHandleLine-69 ResizingControl-resizeHandleLine-113 ResizingControl-resizeHandleSecondLine-115"
                      />
                    </div>
                  </th>
                  <th
                    class="MuiTableCell-root TableHeaderCell-cell-67 TableHeaderCell-cellNoUserSelect-71 TableHeaderCell-cellDraggable-72 TableHeaderCell-cellNoWrap-77 MuiTableCell-head"
                    colspan="1"
                    scope="col"
                  >
                    <div
                      class="TableHeaderCell-container-76"
                    >
                      <div
                        class="Content-content-90"
                      >
                        <div
                          class="SortLabel-root-93"
                        >
                          <span
                            aria-disabled="false"
                            class="MuiButtonBase-root MuiTableSortLabel-root SortLabel-sortLabelRoot-95"
                            role="button"
                            tabindex="0"
                            title="Sort"
                          >
                            <span
                              class="Title-title-111"
                            >
                              Type
                            </span>
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionDesc"
                              focusable="false"
                              role="presentation"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                              />
                            </svg>
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ResizingControl-resizeHandle-112 TableHeaderCell-resizeHandle-68"
                    >
                      <div
                        class="TableHeaderCell-resizeHandleLine-69 ResizingControl-resizeHandleLine-113 ResizingControl-resizeHandleFirstLine-114"
                      />
                      <div
                        class="TableHeaderCell-resizeHandleLine-69 ResizingControl-resizeHandleLine-113 ResizingControl-resizeHandleSecondLine-115"
                      />
                    </div>
                  </th>
                  <th
                    class="MuiTableCell-root TableHeaderCell-cell-67 TableHeaderCell-cellNoUserSelect-71 TableHeaderCell-cellDraggable-72 TableHeaderCell-cellNoWrap-77 MuiTableCell-head"
                    colspan="1"
                    scope="col"
                  >
                    <div
                      class="TableHeaderCell-container-76"
                    >
                      <div
                        class="Content-content-90"
                      >
                        <div
                          class="SortLabel-root-93"
                        >
                          <span
                            aria-disabled="false"
                            class="MuiButtonBase-root MuiTableSortLabel-root SortLabel-sortLabelRoot-95"
                            role="button"
                            tabindex="0"
                            title="Sort"
                          >
                            <span
                              class="Title-title-111"
                            >
                              Sections
                            </span>
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionDesc"
                              focusable="false"
                              role="presentation"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                              />
                            </svg>
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ResizingControl-resizeHandle-112 TableHeaderCell-resizeHandle-68"
                    >
                      <div
                        class="TableHeaderCell-resizeHandleLine-69 ResizingControl-resizeHandleLine-113 ResizingControl-resizeHandleFirstLine-114"
                      />
                      <div
                        class="TableHeaderCell-resizeHandleLine-69 ResizingControl-resizeHandleLine-113 ResizingControl-resizeHandleSecondLine-115"
                      />
                    </div>
                  </th>
                  <th
                    class="MuiTableCell-root TableHeaderCell-cell-67 TableHeaderCell-cellNoUserSelect-71 TableHeaderCell-cellDraggable-72 TableHeaderCell-cellNoWrap-77 MuiTableCell-head"
                    colspan="1"
                    scope="col"
                  >
                    <div
                      class="TableHeaderCell-container-76"
                    >
                      <div
                        class="Content-content-90"
                      >
                        <div
                          class="SortLabel-root-93"
                        >
                          <span
                            aria-disabled="true"
                            class="MuiButtonBase-root MuiTableSortLabel-root SortLabel-sortLabelRoot-95 Mui-disabled"
                            role="button"
                            tabindex="-1"
                            title="Sort"
                          >
                            <span
                              class="Title-title-111"
                            >
                              Action
                            </span>
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionDesc"
                              focusable="false"
                              role="presentation"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                              />
                            </svg>
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ResizingControl-resizeHandle-112 TableHeaderCell-resizeHandle-68"
                    >
                      <div
                        class="TableHeaderCell-resizeHandleLine-69 ResizingControl-resizeHandleLine-113 ResizingControl-resizeHandleFirstLine-114"
                      />
                      <div
                        class="TableHeaderCell-resizeHandleLine-69 ResizingControl-resizeHandleLine-113 ResizingControl-resizeHandleSecondLine-115"
                      />
                    </div>
                  </th>
                  <th
                    class="MuiTableCell-root TableStubCell-cell-118 MuiTableCell-head"
                    colspan="1"
                    scope="col"
                  />
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root TableNoDataCell-cell-121 MuiTableCell-body"
                    colspan="5"
                  >
                    <div
                      class="TableNoDataCell-textContainer-122"
                    >
                      <big
                        class="TableNoDataCell-text-123"
                      >
                        No data
                      </big>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</div>
`;
